### PR TITLE
More informative error when writing attrs to netCDF

### LINKED ISSUE #3080: Error in to_netcdf()
#### MCVE Code Sample

In order for the maintainers to efficiently understand and prioritize issues, we ask you post a "Minimal, Complete and Verifiable Example" (MCVE): http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

```python
import xarray as xr
print(xr.__version__)

url = 'https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/cccs_portal/indices/Final/BCCAQv2/tg_mean/YS/rcp26/simulations/BCCAQv2+ANUSPLIN300_bcc-csm1-1_historical+rcp26_r1i1p1_1950-2100_tg_mean_YS.nc'
ds = xr.open_dataset(url, chunks={'time':50})
dsSel = ds.sel(lon= -80, lat= 50, method='nearest')
outnc = 'tmp.nc'
dsSel.to_netcdf(outnc)

```

#### Problem Description
This may be related to #2850 but I am unable to save to netcdf after performing a simple ```sel``` for certain .nc files (example file available via opendap link in the code) . Gives following error :
```Traceback (most recent call last):
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2961, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-492c82b08e81>", line 8, in <module>
    dsSel.to_netcdf(outnc)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/core/dataset.py", line 1365, in to_netcdf
    compute=compute)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/api.py", line 886, in to_netcdf
    unlimited_dims=unlimited_dims)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/api.py", line 929, in dump_to_store
    unlimited_dims=unlimited_dims)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/common.py", line 269, in store
    self.set_attributes(attributes)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/common.py", line 285, in set_attributes
    self.set_attribute(k, v)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/netCDF4_.py", line 426, in set_attribute
    _set_nc_attribute(self.ds, key, value)
  File "/home/travis/.conda/envs/Xarray/lib/python3.6/site-packages/xarray/backends/netCDF4_.py", line 302, in _set_nc_attribute
    obj.setncattr(key, value)
  File "netCDF4/_netCDF4.pyx", line 2607, in netCDF4._netCDF4.Dataset.setncattr
  File "netCDF4/_netCDF4.pyx", line 1467, in netCDF4._netCDF4._set_att
  File "netCDF4/_netCDF4.pyx", line 1733, in netCDF4._netCDF4._ensure_nc_success
AttributeError: NetCDF: String match to name in use
```

#### Expected Output
saved selection to .nc
#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7 |Anaconda, Inc.| (default, Oct 23 2018, 19:16:44) 
[GCC 7.3.0]
python-bits: 64
OS: Linux
OS-release: 4.15.0-54-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_CA.UTF-8
LOCALE: en_CA.UTF-8
libhdf5: 1.10.2
libnetcdf: 4.6.1
xarray: 0.12.2
pandas: 0.23.3
numpy: 1.15.4
scipy: 1.2.0
netCDF4: 1.4.0
pydap: None
h5netcdf: 0.6.1
h5py: 2.8.0
Nio: None
zarr: None
cftime: 1.0.3.4
nc_time_axis: None
PseudonetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: 1.2.1
dask: 1.0.0
distributed: 1.25.2
matplotlib: 3.0.1
cartopy: None
seaborn: None
numbagg: None
setuptools: 39.2.0
pip: 19.1.1
conda: None
pytest: 3.8.1
IPython: 6.5.0
sphinx: 2.1.2
</details>

### PR DESCRIPTION
**Pull request by fmaussion**
Closes #3080 

Some attribute names aren't valid netCDF4 keys. For example:

```python
 import xarray as xr 
 ds = xr.Dataset({'x': ('y', [1, 2, 3], {'CLASS': 'foo'})})
ds.to_netcdf('test.nc')
```

will fail with:

```NetCDF: String match to name in use```

This is hard to debug if you don't know which attribute is faulty. With this small change the error should be more informative:

```AttributeError: The following exception occurred when attempting to set attribute (CLASS, foo): "NetCDF: String match to name in use"```

### CODE REVIEW

File: xarray/backends/netCDF4_.py [hunk 1]
```diff
@@ -291,7 +291,12 @@ def _set_nc_attribute(obj, key, value):
                    'upgrade to netCDF4-python 1.2.4 or greater.')
             raise AttributeError(msg)
     else:
-        obj.setncattr(key, value)
+        try:
+            obj.setncattr(key, value)
+        except AttributeError as e:
+            msg = ('The following exception occurred when attempting to set '
+                   'attribute ({}, {}): "{}"'.format(key, value, e))
```
Comment (shoyer) [inline xarray/backends/netCDF4_.py]: I don't think we need to include the original error message unless you're silencing the traceback (which is possible). Python 3 will include both tracebacks.

I would suggest just writing something like:
````python
        try:
            obj.setncattr(key, value)
        except AttributeError:
            msg = 'failed to write attribute {} with value {!r} to a netCDF file`.format(key, value)
            raise AttributeError(msg)
```

Comment (dcherian) [reply xarray/backends/netCDF4_.py]: Should we add variable name too, if set?

Comment (shoyer) [reply xarray/backends/netCDF4_.py]: Yes, that would probably be helpful!

**Comment by fmaussion** [comment]
Done! Didn't edit "what's new" because this is a tiny change.

**Comment by dcherian** [comment]
One more request: can you link to a list of disallowed names in the docs somewhere?

I also wonder if the error message should suggest that the user look up this list.

**Comment by fmaussion** [comment]
> One more request: can you link to a list of disallowed names in the docs somewhere?

I don't know if such a list exists. Do you know? It's very python related, because of `setattr`. It's confusing that it's only making a problem on variable objects, not on datasets. 

So altogether I wouldn't worry much here, this is really an edge case.

**Comment by fmaussion** [comment]
> It's very python related

I'm saying this because this happened with a real file I was working with. The file had a `CLASS` attribute on a variable, it can be read by xarray but not written out...

**Comment by shoyer** [comment]
I think this is actually due to something in the netCDF C library: `CLASS` is a special attribute name used by HDF5 dimension scales. Note that it works fine to save an attribute with the name `class` (the Python keyword).

Here's the list of prohibited attributes from the netCDF spec:
https://www.unidata.ucar.edu/software/netcdf/docs/file_format_specifications.html#atts_spec

h5netcdf uses this list to issue more informative error messages:
https://github.com/shoyer/h5netcdf/blob/3ae35cd58297281a1dc69c46fb0b315a0007ac2b/h5netcdf/attrs.py#L22-L24

**Comment by fmaussion** [comment]
Thanks! @shoyer ! I guess the file I was working with must have been created with nectdf3 then, because otherwise I don't know how they managed to write this out...

**Comment by shoyer** [comment]
Yes, you can put these in a netCDF3 file (even with the latest netCDF-C library).

**Pull request closed without merging**